### PR TITLE
[Coverage] Instrument code in primary files only

### PR DESCRIPTION
**Pull request by vedantk**
Unless there are no primary inputs (WMO), code outside of a primary input file
may not be fully typechecked or well-formed. Defer instrumenting non-primary
files with the expectation that that code will be visited later.

This fixes at least one assertion failure seen while building a project
with coverage, and is probably good for some substantial compile-time
improvements with coverage enabled.

rdar://39069115

### CODE REVIEW

**Comment by vedantk** [comment]
+ @slavapestov @AnnaZaks -- heads-up that this might resolve a compile-time issue with code coverage with WMO disabled. I'll take a closer look to confirm.

**Comment by vedantk** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
Thanks!

**Comment by slavapestov** [comment]
Can you add a testcase? We need to make sure that coverage works in non-WMO mode, and also reduce a test for the assertion failure.

File: include/swift/AST/Module.h [hunk 1]
```diff
@@ -830,6 +830,9 @@ class SourceFile final : public FileUnit {
   /// May be -1, to indicate no association with a buffer.
   int BufferID;
 
+  /// \brief Whether this file is sourced from a primary input.
+  bool IsPrimaryInput;
```
Comment (graydon) [inline include/swift/AST/Module.h]: This part (and the general indication of which SourceFiles are primary) ought to be run by @davidungar, he's been trying to centralize reasoning about primary-ness and it'd be a shame if this worked against his efforts.

File: lib/SIL/SILProfiler.cpp [hunk 2]
```diff
@@ -101,6 +101,22 @@ static void walkForProfiling(ASTNode N, ASTWalker &Walker) {
 }
 
 SILProfiler *SILProfiler::create(SILModule &M, ASTNode N) {
+  // Unless there are no primary inputs (WMO), code outside of a primary input
+  // may not be fully formed. Defer instrumenting non-primary files, with the
+  // expectation that that code will be visited later.
+  const auto &Opts = M.getOptions();
```
Comment (graydon) [inline lib/SIL/SILProfiler.cpp]: I don't understand why Opts is moving here.

File: lib/SIL/SILProfiler.cpp [hunk 3]
```diff
@@ -465,21 +480,20 @@ struct PGOMapping : public ASTWalker {
...
-        auto count = loadExecutionCount(elseExpr);
-        if (!parent) {
-          auto thenVal = thenCount.getValue();
-          for (auto pCount = NextCounter - 1; pCount > 0; --pCount) {
-            auto cCount = LoadedCounts->Counts[pCount];
-            if (cCount > thenVal) {
-              count = cCount;
-              break;
-            }
+      auto elseExpr = IE->getElseExpr();
```
Comment (graydon) [inline lib/SIL/SILProfiler.cpp]: I don't understand why this is being un-conditionalized; will there always be a non-null elseExpr now?

**Review by graydon — CHANGES_REQUESTED** [review]

File: include/swift/AST/Module.h (hunk 1 above)
Comment (davidungar) [reply include/swift/AST/Module.h]: I don't yet understand why this is needed. There is a test in FrontendInputsAndOutputs called isInputPrimary. It would be more consistent to use that, but I have just started looking at this PR.
Tx for pinging me, @graydon.

File: lib/Frontend/Frontend.cpp [hunk 4]
```diff
@@ -804,13 +804,14 @@ SourceFile *CompilerInstance::createSourceFileForMainModule(
...
+  bool isPrimary = bufferID && isPrimaryInput(*bufferID);
   SourceFile *inputFile = new (*Context)
       SourceFile(*mainModule, fileKind, bufferID, importKind,
                  Invocation.getLangOptions().CollectParsedToken,
-                 Invocation.getLangOptions().BuildSyntaxTree);
+                 Invocation.getLangOptions().BuildSyntaxTree, isPrimary);
   MainModule->addFile(*inputFile);
 
-  if (bufferID && isPrimaryInput(*bufferID)) {
+  if (isPrimary) {
```
Comment (davidungar) [inline lib/Frontend/Frontend.cpp]: Why is this change needed?

**Review by davidungar — CHANGES_REQUESTED** [review]

File: lib/Frontend/Frontend.cpp (hunk 4 above)
Comment (davidungar) [reply lib/Frontend/Frontend.cpp]: As far as I can tell,  this is creating redundant information about primaryness. Is there some reason to redundantly store the boolean in the SourceFile?

Comment (slavapestov) [reply lib/Frontend/Frontend.cpp]: I think the crux of the issue is that we're creating a SILProfiler per SILFunction, and SILFunctions are used for declarations as well as definitions. For a declaration, we should not be creating a SILProfiler, nor should we be walking the AST of the function.

File: include/swift/AST/Module.h (hunk 1 above)
Comment (jrose-apple) [reply include/swift/AST/Module.h]: Yes, please do not put information about primary-ness into SourceFile. It's the responsibility of clients to not try to do things to non-primary SourceFiles, and most of the compiler is set up so that that can't happen by accident.

Comment (vedantk) [reply include/swift/AST/Module.h]: Thanks for your feedback. I wanted to add an isPrimary predicate to SourceFile because I thought I needed some way to get that information at the SIL level. As Jordan and Slava pointed out, there should be a simpler approach.

File: lib/SIL/SILProfiler.cpp (hunk 2 above)
Comment (vedantk) [reply lib/SIL/SILProfiler.cpp]: This was accidental code movement, sorry about that.

File: lib/SIL/SILProfiler.cpp (hunk 3 above)
Comment (vedantk) [reply lib/SIL/SILProfiler.cpp]: Given a well-formed AST, the else subexpression of an IfExpr should always be non-null. The underlying bug here is that the coverage pass goes off and operates on incomplete ASTs.

File: lib/Frontend/Frontend.cpp (hunk 4 above)
Comment (vedantk) [reply lib/Frontend/Frontend.cpp]: Thanks, I'll look into this!

**Comment by vedantk** [comment]
Thanks for your comments, everyone. I've implemented a simpler solution (per Slava's suggestion) in https://github.com/apple/swift/pull/15773. PTAL.
